### ISSUE TITLE
Expanding certificates with Apache or Nginx installers consistently crap out w/regards to rewrites. Bugs or lack of documentation?

### ISSUE THREAD

**Issue by Kramerican** [certbot/certbot#5615]
## My operating system is (include version):
Ubuntu Xenial 16.04

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
apt (certbot)

## Certbot's behavior differed from what I expected because:

Re-running Certbot in order to expand certificates while using --installer apache or --installer nginx and using --redirect consistently blows up.

It seems to me that this should really "just work" and I am baffled why it isn't. Seeing that so many other hosting companies are using LetsEncrypt - and I assumed leveraged Certbot - I'd think this would be a very common workflow which was rock-solid at this point.

This leads me to think that A) either hosting companies in general are not using Certbot, or B) I am using Certbot wrong.

## To reproduce (Apache):

Generate certs and use --installer apache on a clean Virtualhost where Servername (test.domain.com) is defined in the vhost config:

`certbot --webroot-path /var/www/html --authenticator webroot --installer apache --expand --non-interactive --agree-tos --email=my@mail.com --domains test.domain.com --redirect`

Now point another domain to the server, update the http vhost config and add it as a ServerAlias, and re-run Certbot with near-identical cmd line:

`certbot --webroot-path /var/www/html --authenticator webroot --installer apache --expand --non-interactive --agree-tos --email=my@mail.com --domains myrealdomain.com,test.domain.com --redirect`

**Result**

Certbot will fail with errors relating to being unable to expand/update rewrite rules and state **Unable to install the certificate**

This to me is exceedingly bizarre. Certbot made the rewrite rules in the first place, and being unable to modify these is a catastrophe.

## To reproduce (Nginx):

Follow the exact same workflow as the above, of course defining the domains in the vhost config before runinng Certbot. 

**Result**

Certbot will now fail authenticating the domain using webroot as it gets a 404 not found. Why? Because it has in the first run added this to the Nginx config:

```
server {
    if ($host = test.domain.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

        listen 80 default_server;
        listen [::]:80 default_server;
    return 404; # managed by Certbot
}
```
I can almost see in the crystal ball that you'll just say here: Stop using webroot auth you numbnut! Well ... I am a host, and I don't know beforehand whether users have proxied their domain through e.g. Cloudflare. In this scenario webroot is the only auth method that works. So if I want to script my way out of that ... Well.. That would entail me having to try and look up the domain and see if it points directly to the server IP or not.

Yes, that would be a solution, but in my world Certbot made this vhost config, so it should really be able to deal with this? Maybe I'm being unreasonable here.

Besides, I am almost certain (did not test this, this time around!) that Certbot would still fail due to perceived ambiguity regarding redirects.

## Shouldn't Certbot be able to handle ServerName/ServerAlias (Apache) and server_name (Nginx) ?

It seems to me that in an ideal world I would never have to touch the server config. Certbot should be able to figure out that I am defining a new set of domains for this server (or expanding on the old set) and set vhost config appropriately and handle rewrites gracefully.

I have of course tested this workflow where I haven't set ServerName/Alias - and then I of course get a failure as Certbot says that it really wants me to define my domains in my vhost config. So yeahh...

## All this assumes that this is something Certbot is designed to do. It seems to me it is not.

Is the recommended workflow, in fact, when you want to expand certficiates, and use --installer and --redirect, is to:

1. Execute Certbot rollback
2. Do your vhost updates (updating domain names)
3. Re-run Certbot

This is the only approach I have gotten to consistently work and not blow up due to Certbot choking on its own rewrite config.

**There is a downside here however: If the user has done any customization in the meantime, the Certbot rollback will nuke all their changes.**

## If this is something Certbot should be able to do, then:

1. You should really test for this. It's not a complicated workflow.
2. Maybe consider having Certbot handle domain name definitions in the vhost config files (yes, I know this is exceedingly complicated)

**Comment by joohoi** [comment]
Thanks for the report! I'm unable to reproduce the problem against the current master (I tested with Apache). Which version of Certbot are you using? It would also be useful to see the contents of `/var/log/letsencrypt/letsencrypt.log` for the failed run to see where it crashes.

**Comment by Kramerican** [comment]
@joohoi 

I am using the latest Certbot, **certbot 0.21.1**

**First of all:**

You are not willing to comment on my overall questions? Am I to take it that my workflow is fundamentally correct and that Certbot should be able to handle this?

I am frankly extremely surprised you cannot reproduce this, as I can reproduce this quite reliably in both Apache and Nginx. Yes, it is differing issues for each webserver, but Certbot breaks quite easily with both installers in my experience.

In any case, what I am battling right now, in the case of Apache is the following. I am working on the rollback -> set vhost names -> re-run Certbot workflow at the moment, so this is what I can report right now. I can go back and test the "Just run Certbot without rollback" in a moment, but that breaks as well in a similar fashion. 

1. Start with a vhost like so

<details><summary>CLICK ME</summary>
<p>

```
<VirtualHost *:80>
ServerName testkrellide.krellide.webdock.io

        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Deny access to any git repository
        RedirectMatch 404 /\.git

    <FilesMatch \.php$>
        # 2.4.10+ can proxy to unix socket
         SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>

    <Directory /var/www/>
        Options  FollowSymLinks Includes ExecCGI
        AllowOverride All
        Require all granted
    </Directory>

        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel alert rewrite:trace3 alias:debug

        CustomLog /var/www/logs/access.log combined
        ErrorLog /var/www/logs/error.log

</VirtualHost>

<Directory /usr/share/phpMyAdmin>
Options FollowSymlinks
AllowOverride all
Require all granted
</Directory>
```

</p>
</details>


2. Now run Certbot 

`certbot --webroot-path /var/www/html --authenticator webroot --installer apache --expand --non-interactive --agree-tos --email=my@mail.com --domains testkrellide.krellide.webdock.io --redirect`

Everything is fine. Now I revert the config

`certbot rollback`

3. And update the vhost

<details><summary>CLICK ME</summary>
<p>

```
<VirtualHost *:80>
ServerName testingnew.webdock.io
ServerAlias testkrellide.krellide.webdock.io

        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Deny access to any git repository
        RedirectMatch 404 /\.git

    <FilesMatch \.php$>
        # 2.4.10+ can proxy to unix socket
         SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>

    <Directory /var/www/>
        Options  FollowSymLinks Includes ExecCGI
        AllowOverride All
        Require all granted
    </Directory>

        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel alert rewrite:trace3 alias:debug

        CustomLog /var/www/logs/access.log combined
        ErrorLog /var/www/logs/error.log

</VirtualHost>

<Directory /usr/share/phpMyAdmin>
Options FollowSymlinks
AllowOverride all
Require all granted
</Directory>
```
</p>
</details>

4. Re-run Certbot

`certbot --webroot-path /var/www/html --authenticator webroot --installer apache --expand --non-interactive --agree-tos --email=my@mail.com --domains testingnew.webdock.io,testkrellide.krellide.webdock.io --redirect`

I get the error output:

<details><summary>CLICK ME</summary>
<p>

```
Certbot failed in testkrellide Certbot message: Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer apache
Renewing an existing certificate
Performing the following challenges:
http-01 challenge for testingnew.webdock.io
http-01 challenge for testkrellide.krellide.webdock.io
Using the webroot path /var/www/html for all unmatched domains.
Waiting for verification...
Cleaning up challenges
Created an SSL vhost at /etc/apache2/sites-available/webdock-le-ssl.conf
Deploying Certificate for testingnew.webdock.io to VirtualHost /etc/apache2/sites-available/webdock-le-ssl.conf
An unexpected error occurred:
ValueError: Unable to set value to path!
Please see the logfiles in /var/log/letsencrypt for more details.
IMPORTANT NOTES:
 - Unable to install the certificate
 - Congratulations! Your certificate and chain have been saved at:
   /etc/letsencrypt/live/testkrellide.krellide.webdock.io/fullchain.pem
   Your key file has been saved at:
   /etc/letsencrypt/live/testkrellide.krellide.webdock.io/privkey.pem
   Your cert will expire on 2018-05-24. To obtain a new or tweaked
   version of this certificate in the future, simply run certbot again
   with the "certonly" option. To non-interactively renew *all* of
   your certificates, run "certbot renew"
```

</p>
</details>

And /var/log/letsencrypt/letsencrypt.log reports

<details><summary>CLICK ME</summary>
<p>


```
2018-02-23 19:07:15,818:DEBUG:certbot.reporter:Reporting to user: Congratulations! Your certificate and chain have been saved at:
/etc/letsencrypt/live/testkrellide.krellide.webdock.io/fullchain.pem
Your key file has been saved at:
/etc/letsencrypt/live/testkrellide.krellide.webdock.io/privkey.pem
Your cert will expire on 2018-05-24. To obtain a new or tweaked version of this certificate in the future, simply run certbot again with the "certonly" option. To non-interactively renew *all* of your certificates, run "certbot renew"
2018-02-23 19:07:15,818:DEBUG:certbot.reverter:Creating backup of /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:07:15,828:INFO:certbot_apache.configurator:Created an SSL vhost at /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:07:15,982:INFO:certbot_apache.configurator:Deploying Certificate for testingnew.webdock.io to VirtualHost /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:07:16,033:DEBUG:certbot.error_handler:Encountered exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 452, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 292, in deploy_cert
    self._add_dummy_ssl_directives(vhost.path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 1269, in _add_dummy_ssl_directives
    "insert_cert_file_path")
  File "/usr/lib/python3/dist-packages/certbot_apache/parser.py", line 327, in add_dir
    self.aug.set(aug_conf_path + "/directive[last() + 1]", directive)
  File "/usr/lib/python3/dist-packages/augeas.py", line 187, in set
    raise ValueError("Unable to set value to path!")
ValueError: Unable to set value to path!

2018-02-23 19:07:16,033:DEBUG:certbot.error_handler:Calling registered functions
2018-02-23 19:07:16,036:DEBUG:certbot.reporter:Reporting to user: Unable to install the certificate
2018-02-23 19:07:16,037:DEBUG:certbot.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/certbot", line 11, in <module>
    load_entry_point('certbot==0.21.1', 'console_scripts', 'certbot')()
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1240, in main
    return config.func(config, plugins)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1001, in run
    _install_cert(config, le_client, domains, new_lineage)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 757, in _install_cert
    path_provider.cert_path, path_provider.chain_path, path_provider.fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 452, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 292, in deploy_cert
    self._add_dummy_ssl_directives(vhost.path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 1269, in _add_dummy_ssl_directives
    "insert_cert_file_path")
  File "/usr/lib/python3/dist-packages/certbot_apache/parser.py", line 327, in add_dir
    self.aug.set(aug_conf_path + "/directive[last() + 1]", directive)
  File "/usr/lib/python3/dist-packages/augeas.py", line 187, in set
    raise ValueError("Unable to set value to path!")
ValueError: Unable to set value to path!
```

</p>
</details>

**Comment by Kramerican** [comment]
@joohoi 

Doing the exact same workflow, and this time not doing `certbot rollback` but only touching the vhost config in the lightest way possible (adding the domain) so it becomes:

<details><summary>CLICK ME</summary>
<p>

```
<VirtualHost *:80>
ServerName testingnew.webdock.io
ServerAlias testkrellide.krellide.webdock.io

        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Deny access to any git repository
        RedirectMatch 404 /\.git

    <FilesMatch \.php$>
        # 2.4.10+ can proxy to unix socket
         SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>

    <Directory /var/www/>
        Options  FollowSymLinks Includes ExecCGI
        AllowOverride All
        Require all granted
    </Directory>

        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel alert rewrite:trace3 alias:debug

        CustomLog /var/www/logs/access.log combined
        ErrorLog /var/www/logs/error.log

RewriteEngine on
RewriteCond %{SERVER_NAME} =testkrellide.krellide.webdock.io
RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>

<Directory /usr/share/phpMyAdmin>
Options FollowSymlinks
AllowOverride all
Require all granted
</Directory>

```

</p>
</details>

Results in the same error:

<details><summary>CLICK ME</summary>
<p>

```
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer apache
Renewing an existing certificate
Performing the following challenges:
http-01 challenge for testingnew.webdock.io
http-01 challenge for testkrellide.krellide.webdock.io
Using the webroot path /var/www/html for all unmatched domains.
Waiting for verification...
Cleaning up challenges
Created an SSL vhost at /etc/apache2/sites-available/webdock-le-ssl.conf
Deploying Certificate for testingnew.webdock.io to VirtualHost /etc/apache2/sites-available/webdock-le-ssl.conf
An unexpected error occurred:
ValueError: Unable to set value to path!
Please see the logfiles in /var/log/letsencrypt for more details.

IMPORTANT NOTES:
 - Unable to install the certificate
 - Congratulations! Your certificate and chain have been saved at:
   /etc/letsencrypt/live/testkrellide.krellide.webdock.io/fullchain.pem
   Your key file has been saved at:
   /etc/letsencrypt/live/testkrellide.krellide.webdock.io/privkey.pem
   Your cert will expire on 2018-05-24. To obtain a new or tweaked
   version of this certificate in the future, simply run certbot again
   with the "certonly" option. To non-interactively renew *all* of
   your certificates, run "certbot renew"
 - Some rewrite rules copied from
   /etc/apache2/sites-enabled/webdock.conf were disabled in the vhost
   for your HTTPS site located at
   /etc/apache2/sites-available/webdock-le-ssl.conf because they have
   the potential to create redirection loops.

```

</p>
</details>

And the log:

<details><summary>CLICK ME</summary>
<p>

```
2018-02-23 19:28:44,120:DEBUG:certbot.storage:Writing new config /etc/letsencrypt/renewal/testkrellide.krellide.webdock.io.conf.new.
2018-02-23 19:28:44,122:DEBUG:certbot.reporter:Reporting to user: Congratulations! Your certificate and chain have been saved at:
/etc/letsencrypt/live/testkrellide.krellide.webdock.io/fullchain.pem
Your key file has been saved at:
/etc/letsencrypt/live/testkrellide.krellide.webdock.io/privkey.pem
Your cert will expire on 2018-05-24. To obtain a new or tweaked version of this certificate in the future, simply run certbot again with the "certonly" option. To non-interactively renew *all* of your certificates, run "certbot renew"
2018-02-23 19:28:44,122:DEBUG:certbot.reverter:Creating backup of /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:28:44,128:DEBUG:certbot.reporter:Reporting to user: Some rewrite rules copied from /etc/apache2/sites-enabled/webdock.conf were disabled in the vhost for your HTTPS site located at /etc/apache2/sites-available/webdock-le-ssl.conf because they have the potential to create redirection loops.
2018-02-23 19:28:44,132:INFO:certbot_apache.configurator:Created an SSL vhost at /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:28:44,292:INFO:certbot_apache.configurator:Deploying Certificate for testingnew.webdock.io to VirtualHost /etc/apache2/sites-available/webdock-le-ssl.conf
2018-02-23 19:28:44,343:DEBUG:certbot.error_handler:Encountered exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 452, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 292, in deploy_cert
    self._add_dummy_ssl_directives(vhost.path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 1269, in _add_dummy_ssl_directives
    "insert_cert_file_path")
  File "/usr/lib/python3/dist-packages/certbot_apache/parser.py", line 327, in add_dir
    self.aug.set(aug_conf_path + "/directive[last() + 1]", directive)
  File "/usr/lib/python3/dist-packages/augeas.py", line 187, in set
    raise ValueError("Unable to set value to path!")
ValueError: Unable to set value to path!

2018-02-23 19:28:44,343:DEBUG:certbot.error_handler:Calling registered functions
2018-02-23 19:28:44,346:DEBUG:certbot.reporter:Reporting to user: Unable to install the certificate
2018-02-23 19:28:44,347:DEBUG:certbot.log:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/certbot", line 11, in <module>
    load_entry_point('certbot==0.21.1', 'console_scripts', 'certbot')()
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1240, in main
    return config.func(config, plugins)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 1001, in run
    _install_cert(config, le_client, domains, new_lineage)
  File "/usr/lib/python3/dist-packages/certbot/main.py", line 757, in _install_cert
    path_provider.cert_path, path_provider.chain_path, path_provider.fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot/client.py", line 452, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 292, in deploy_cert
    self._add_dummy_ssl_directives(vhost.path)
  File "/usr/lib/python3/dist-packages/certbot_apache/configurator.py", line 1269, in _add_dummy_ssl_directives
    "insert_cert_file_path")
  File "/usr/lib/python3/dist-packages/certbot_apache/parser.py", line 327, in add_dir
    self.aug.set(aug_conf_path + "/directive[last() + 1]", directive)
  File "/usr/lib/python3/dist-packages/augeas.py", line 187, in set
    raise ValueError("Unable to set value to path!")
ValueError: Unable to set value to path!

```

</p>
</details>

However, this is something new. I did not have this specific error before, but rather "just" had the errors also reported later in the output - **please note the messages regarding the rewrite rules causing redirection loops near the end of the output** ...This was my "classic" issue.

**Comment by Kramerican** [comment]
With the above in mind:

Seems to me possibly I've uncovered a bug which wasn't there before. This is the first time I am seeing this path value stuff in anyway.

BUT

**My original issue question still stands:** What is the correct workflow here? Should you do rollback? If not, should you also remove or alter redirects by hand before re-running Certbot? Why isn't this documented in a how-to?

**Comment by joohoi** [comment]
This is lot to go through. I'll prepare a comprehensive answer for you over the weekend, but I'll just quickly describe the `rollback` and `expand` here, as I feel this workflow is suboptimal.

If there's a need to add new domains to an already existing certificate, `--expand` is to be used. This however is rarely the case, and generally users prefer to have more certificates with smaller subsets of domain, or even a separate certificate per domain. This works with the same command line you are using, only leaving `--expand` and the already existing domains off.

Rollback is used when you want to revert back to your original configuration, discarding all the changes made by Certbot. This is used in the error situations usually, and I haven't seen it used in an actual workflow before. Leaving rollback out should work for you.

I'll prepare a more comprehensive answer for you later, but these are my initial thoughts.

**Comment by Kramerican** [comment]
@joohoi 

As I suspected, my workflow sucks :) ... Thank you for the very illumating answer! I will dig into this some more and try and find a workflow that works for us.

My immediate comment on your notes re. omitting --expand is that **if** we are to only run Certbot indicating new domains - i.e. leaving existing domains off - this then requires me to write up code to keep track of state changes.

i.e. I would then need to know in my scripts which domains had certs generated already, and which didn't. This is something which can get hairy and buggy quite fast I suspect, so that's why I had opted to do the --expand workflow, as that didn't require me to keep track of anything. I could just re-issue the command with my full list of domains and it would... Well, break... Haha :)

Any thoughts on this particular aspect?

**Comment by joohoi** [comment]
> Any thoughts on this particular aspect?

Technically in the background expanding an existing certificate is equivalent to requesting a new one. So what you could do in your script magic, could be to use some dark regexp magic to extract `ServerName` and `ServerAlias`es off the `VirtualHost` you want to configure, and feed them all with the newcomer to the Certbot command line. This wouldn't require you to save a state beforehand, but would be a step towards more dynamic way of handling the issue.

**Comment by Kramerican** [comment]
Ok thanks

Hmm - Dark magic doesn't sit well with me. Besides, I suspect this is not a lifesaver here, as I Just tried the exact same workflow as before (squeaky clean system), but doing as you suggested and omitting --expand and just passing the new domain. This succeeds somewhat, but uncovers my original pain point: --redirect doesn't work as expected. What I get is:

```
...
Using the webroot path /var/www/html for all unmatched domains.
Waiting for verification...
Cleaning up challenges
Created an SSL vhost at /etc/apache2/sites-available/webdock-le-ssl.conf
Deploying Certificate for testingnew.webdock.io to VirtualHost /etc/apache2/sites-available/webdock-le-ssl.conf
Enhancement redirect was already set.
...
```

which basically means redirects were not added to my original vhost config file for the new domain - and I end up with a right mess of a vhost file generated by Certbot (/etc/apache2/sites-enabled/webdock-le-ssl.conf)

<details><summary>CLICK ME</summary>
<p>

```
<IfModule mod_ssl.c>
<VirtualHost *:443>
ServerName testkrellide.krellide.webdock.io

        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Deny access to any git repository
        RedirectMatch 404 /\.git

    <FilesMatch \.php$>
        # 2.4.10+ can proxy to unix socket
         SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>

    <Directory /var/www/>
        Options  FollowSymLinks Includes ExecCGI
        AllowOverride All
        Require all granted
    </Directory>

        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel alert rewrite:trace3 alias:debug

        CustomLog /var/www/logs/access.log combined
        ErrorLog /var/www/logs/error.log

Include /etc/letsencrypt/options-ssl-apache.conf
ServerAlias testingnew.webdock.io
SSLCertificateFile /etc/letsencrypt/live/testingnew.webdock.io/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/testingnew.webdock.io/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
</IfModule>
<IfModule mod_ssl.c>
<VirtualHost *:80>
ServerName testingnew.webdock.io
ServerAlias testkrellide.krellide.webdock.io

        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        # Deny access to any git repository
        RedirectMatch 404 /\.git

    <FilesMatch \.php$>
        # 2.4.10+ can proxy to unix socket
         SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>

    <Directory /var/www/>
        Options  FollowSymLinks Includes ExecCGI
        AllowOverride All
        Require all granted
    </Directory>

        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel alert rewrite:trace3 alias:debug

        CustomLog /var/www/logs/access.log combined
        ErrorLog /var/www/logs/error.log

RewriteEngine on
# Some rewrite rules in this file were disabled on your HTTPS site,
# because they have the potential to create redirection loops.

# RewriteCond %{SERVER_NAME} =testkrellide.krellide.webdock.io
# RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>
</IfModule>


```

</p>
</details>


I mean... Holy crap! And now the original domain does not work.

When you go through all this in the weekend, I would LOVE to hear your thoughts on the appropriate workflow for running certbot where:

1. Users can add or remove domains at any time and then run Certbot
- How much should I touch the vhost config? Should I be creating seperate vhost files per domain, or what are we looking at here?
- What are the differing workflows here for Apache and Nginx?

2. Auth method webroot is preferred as users may have proxied domains
- But is there a way to tell Certbot to do fallback auth? Meaning something like "Try default, if that doesn't work try webroot" ?

3. --redirect is a must
- But seems to be the root of all evil. What to do??

**Comment by joohoi** [comment]
I would recommend posting these questions to the [Let's Encrypt community forums](https://community.letsencrypt.org/). There's a lot of skilled people familiar with all kinds of workflows there, and you will most likely get better ideas from there. I'll try to answer your questions later though.

**Comment by Kramerican** [comment]
Thanks @joohoi 

I tried doing that, and got the general response: "Certbot should be able to handle it, you're probably running into bugs, go post in the issue tracker" :laughing: 

I'll await your answers, and on the basis of that I might then go back to the forums and ask again, in a different way based on what we learn here.

**Comment by Kramerican** [comment]
As a sidenote, to round all this off on this fine friday night: :)

I have now found the issue with regards to the original bug in my very first post where I was doing rollbacks: **For some reason issuing certbot rollback does in fact NOT roll back the vhost config!**

This meant that the original Certbot SSL config for Apache was the cause of the ambiguity - I got hints from this post which led me to inspect closer the system state before re-running Certbot: [LE Community](https://community.letsencrypt.org/t/valueerror-during-aws-ami-linux-certbot-install/38038/4)

In any case, I have to run `certbot rollback` twice to return to the original state. After which everything runs as it should.

So what I am doing now is the rollback -> rollback -> set names -> run certbot on every run with --expand. This has the aforementioned downside of nuking any custom config the user might have made in the meantime. But at least it works! And now I can generate certs when users provision servers, for that awesome LE experience ;O)

As a fledgling hosting company which just wants to give our users the best possible experience, and given that the official mission statement is to encourage LetsEncrypt adoption as widely as possibe - and seeing that what you **really want, it seems** is for us hosts to secure domains by default - then I must say that at this point it is unclear how we are supposed to achieve this. In an ideal world, the following should be 100% supported by Certbot across both major webservers:

1. The ability to run Certbot with any arbitrary list of domains, and where Certbot will figure out how to update vhost files correctly on its own

2. Certbot should not poop its pants when it comes to redirects. Especially if it's redirect blocks **it itself has generated in the first place**

3. Certbot should ideally be smart about auth and be able to fallback to webroot from the default method if specified

4. And finally, documentation/guidelines on how a webhost, like us, which sets up clean virtual machines with standard vhost config should invoke Certbot for best results

@joohoi THANK YOU for all your time and attention on this..! I will stick with my current approach for now, until we can figure out something better together :)
